src/Adapter/Product/Stock/Update/ProductStockUpdater.php

```diff
@@ -190,40 +190,37 @@ private function updateStockAvailable(StockAvailable $stockAvailable, ProductSto
             $stockUpdateRequired = true;
         }
 
-        if (null !== $properties->getDeltaQuantity()) {
-            $this->updateQuantity($stockAvailable, $properties->getDeltaQuantity());
+        if ($properties->getDeltaQuantity()) {
+            $stockAvailable->quantity += $properties->getDeltaQuantity()->getDeltaQuantity();
             $stockUpdateRequired = true;
         }
 
-        if ($stockUpdateRequired) {
-            $this->stockAvailableRepository->update($stockAvailable);
+        if (!$stockUpdateRequired) {
+            return;
+        }
+
+        $this->stockAvailableRepository->update($stockAvailable);
+
+        if ($properties->getDeltaQuantity()) {
+            //Save movement only after stock has been updated
+            $this->saveMovement($stockAvailable, $properties->getDeltaQuantity());
         }
     }
 
     /**
      * @param StockAvailable $stockAvailable
-     * @param int $deltaQuantity
+     * @param DeltaQuantity $deltaQuantity
      */
-    private function updateQuantity(StockAvailable $stockAvailable, int $deltaQuantity): void
+    private function saveMovement(StockAvailable $stockAvailable, DeltaQuantity $deltaQuantity): void
     {
-        $stockAvailable->quantity += $deltaQuantity;
-
-        if (0 !== $deltaQuantity) {
-            if ($deltaQuantity > 0) {
-                $movementReasonId = $this->configuration->getInt('PS_STOCK_MVT_INC_EMPLOYEE_EDITION');
-            } else {
-                $movementReasonId = $this->configuration->getInt('PS_STOCK_MVT_DEC_EMPLOYEE_EDITION');
-            }
-
-            $this->stockManager->saveMovement(
-                $stockAvailable->id_product,
-                $stockAvailable->id_product_attribute,
-                $deltaQuantity,
-                [
-                    'id_stock_mvt_reason' => $movementReasonId,
-                ]
-            );
-        }
+        $this->stockManager->saveMovement(
```

Comment:
Same suggestion here as in combination updater we need to be sure that the delta is not zero:
- either in the VO constructor by throwing an exception when 0 is used (then the check is performed in the builder to avoid an error)
- or in each updater service